core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts

```diff
@@ -0,0 +1,214 @@
+ *  3. when dragging ends, the temporary DOM element ("flyPaper") is destoryed by jQueryUI
+ *  4. when dragging ends (operator is dropped), it will notify the observer of the event,
+ *    the Operator UI Serivce is responsible for creating an operator at the place dropped.
+ *
+ * The method mentioned above is the best working way to implement this functionailtiy as of 02/2018.
+ * Here are some other methods that have been tried but didn't work.
+ *
+ *  1. Using HTML5 native drag and drop API.
+ *    This doesn't work because the problem of the "ghost image" (the image that follows the mouse when dragging).
+ *    The native HTML5 drag/drop API requires that the "ghost" image must be visually the same as the original element.
+ *    However, in our case, the dragging operator is not the same as the original element.
+ *    There is **NO** workaround for this problem: see this post for details: https://kryogenix.org/code/browser/custom-drag-image.html
+ *      (part of the post isn't exactly true on Chrome anymore because the Chrome itself changed)
+ *    The HTML5 drag and Drop API itself is also considered a disaster: https://www.quirksmode.org/blog/archives/2009/09/the_html5_drag.html
+ *
+ *  2. Using some angular drag and drop libraries for Angular, for example:
+ *    ng2-dnd: https://github.com/akserg/ng2-dnd
+ *    ng2-dragula: https://github.com/valor-software/ng2-dragula
+ *    ng-drag-drop: https://github.com/ObaidUrRehman/ng-drag-drop
+ *
+ *    These drag and drop libraries have the same ghost image problem mentioned above. Moreover, some of them are designed
+ *      for moving a DOM element to another place by dragging and dropping, which is not we want.
+ *
+ * @author Zuozhi Wang
+ *
+ */
+@Injectable()
+export class DragDropService {
+
+  private static readonly DRAG_DROP_TEMP_OPERATOR_TYPE = 'drag-drop-temp-operator-type';
+
+  /** mapping of DOM Element ID to operatorType */
+  private elementOperatorTypeMap = new Map<string, string>();
+  /** the current operatorType of the operator being dragged */
+  private currentOperatorType = DragDropService.DRAG_DROP_TEMP_OPERATOR_TYPE;
+
+
+  /** Subject for operator dragging is started */
+  private operatorDragStartedSubject = new Subject<{ operatorType: string }>();
+
+  /** Subject for operator is dropped on the main workflow editor (equivalent to dragging is stopped) */
+  private operatorDroppedSubject = new Subject<{
+    operatorType: string,
+    offset: Point
+  }>();
+
+  constructor(
+    private jointUIService: JointUIService,
+    private workflowUtilService: WorkflowUtilService,
+    private workflowActionService: WorkflowActionService
+  ) {
+    this.handleOperatorDropEvent();
+  }
+
+  public handleOperatorDropEvent(): void {
+    this.getOperatorDropStream().subscribe(
+      value => {
+        this.currentOperatorType = DragDropService.DRAG_DROP_TEMP_OPERATOR_TYPE;
```

Comment:
Is this like resetting the value to some non-existent type?

Reply:
yes